GitUI/GitUIExtensions.cs

```diff
@@ -34,82 +34,90 @@ public static class GitUIExtensions
+                return fileViewer.ViewGitItemRevisionAsync(file, selectedRev.ObjectId, openWithDiffTool);
             }
 
-            return patch.Text;
-        }
+            string selectedPatch = GetSelectedPatch(fileViewer, firstId, selectedRev.ObjectId, file);
+            return fileViewer.ViewPatchAsync(file.Name, text: selectedPatch ?? defaultText,
+                openWithDifftool: openWithDiffTool, isText: file.IsSubmodule || selectedPatch == null);
 
-        public static Task ViewChangesAsync(this FileViewer diffViewer,
-            [CanBeNull] ObjectId firstId,
-            [CanBeNull] GitRevision selectedRevision,
-            [NotNull] GitItemStatus file,
-            [NotNull] string defaultText,
-            [CanBeNull] Action openWithDifftool = null)
-        {
-            if (firstId == null && selectedRevision != null)
+            void OpenWithDiffTool()
             {
-                firstId = selectedRevision.FirstParentGuid;
+                fileViewer.Module.OpenWithDifftool(
+                    file.Name,
+                    file.OldName,
+                    firstId?.ToString(),
+                    selectedRev?.ToString(),
+                    "",
+                    file.IsTracked);
             }
 
-            openWithDifftool ??= OpenWithDifftool;
-            if (file.IsNew || firstId == null || FileHelper.IsImage(file.Name))
+            static string GetSelectedPatch(
+                FileViewer fileViewer,
+                ObjectId firstId,
+                ObjectId selectedId,
+                GitItemStatus file)
             {
-                // The previous commit does not exist, nothing to compare with
-                if (selectedRevision == null)
+                if (firstId == ObjectId.CombinedDiffId)
                 {
-                    throw new ArgumentNullException(nameof(selectedRevision));
+                    var diffOfConflict = fileViewer.Module.GetCombinedDiffContent(selectedId, file.Name,
+                        fileViewer.GetExtraDiffArguments(), fileViewer.Encoding);
+
+                    return string.IsNullOrWhiteSpace(diffOfConflict)
+                        ? diffOfConflict
+                        : Strings.UninterestingDiffOmitted;
                 }
 
-                // View blob guid from revision, or file for worktree
-                return diffViewer.ViewGitItemRevisionAsync(file, selectedRevision.ObjectId, openWithDifftool);
-            }
+                if (file.IsSubmodule && file.GetSubmoduleStatusAsync() != null)
+                {
+                    // Patch already evaluated
+                    var status = ThreadHelper.JoinableTaskFactory.Run(file.GetSubmoduleStatusAsync);
+                    return status != null
+                        ? LocalizationHelpers.ProcessSubmoduleStatus(fileViewer.Module, status)
+                        : $"Submodule \"{file.Name}\" has unresolved conflicts";
```

Comment:
Consider localization. Perhaps rather support `null` in `ProcessSubmoduleStatus`.

Reply:
I cannot see how the submodule status would not be evaluated except in case of errors, so I changed the message and suggest no translation.